lint/rule.go

```diff
@@ -14,10 +14,295 @@
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *FileRule) GetURI() string {
+	return r.URI
+}
+
+// Lint forwards the FileDescriptor to the LintFile method defined on the
+// FileRule.
+func (r *FileRule) Lint(fd *desc.FileDescriptor) []Problem {
+	return r.LintFile(fd)
+}
+
+// MessageRule defines a lint rule that is run on each message (top-level or
+// nested) within a file.
+type MessageRule struct {
+	Name RuleName
+	URI  string
+
+	// LintMessage accepts a MessageDescriptor and lints it, returning a slice
+	// of Problems it finds.
+	LintMessage func(*desc.MessageDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *MessageRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *MessageRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and iterates over every message in the
+// file, and lints each message in the file.
+func (r *MessageRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+
+	// Iterate over each message and process rules for each message.
+	for _, message := range getAllMessages(fd) {
+		problems = append(problems, r.LintMessage(message)...)
+	}
+	return problems
+}
+
+// FieldRule defines a lint rule that is run on each field within a file.
+type FieldRule struct {
+	Name RuleName
+	URI  string
+
+	// LintField accepts a FieldDescriptor and lints it, returning a slice of
+	// Problems it finds.
+	LintField func(*desc.FieldDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *FieldRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *FieldRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and lints every field in the file.
+func (r *FieldRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+
+	// Iterate over each message and process rules for each field in that
+	// message.
+	for _, message := range getAllMessages(fd) {
+		for _, field := range message.GetFields() {
+			problems = append(problems, r.LintField(field)...)
+		}
+	}
+	return problems
+}
+
+// ServiceRule defines a lint rule that is run on each service.
+type ServiceRule struct {
+	Name RuleName
+	URI  string
+
+	// LintService accepts a ServiceDescriptor and lints it.
+	LintService func(*desc.ServiceDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *ServiceRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *ServiceRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and lints every service in the file.
+func (r *ServiceRule) Lint(fd *desc.FileDescriptor) (problems []Problem) {
```

Comment:
Please don't use named result parameters, which will stutter in godoc.

Reply:
Changed to non-named parameters, albeit for different reasons (`reflect.DeepEqual` barfs when we return `nil`).